# Comprensión del negocio
 
- **Contexto**
 
La empresa opera una plataforma de comercio electrónico que registra sus transacciones como lo son compra, pago, envío y experiencia del cliente desde 2022 hasta 2024. Actualmente esta información se almacena pero no se explota de forma sistemática para entender el comportamiento de compra ni anticipar problemas como las devoluciones.
 
- **Objetivo de negocio**
 
¿Cómo pueden los datos históricos de transacciones en una plataforma de comercio electrónico utilizarse para identificar patrones de compra, predecir devoluciones de productos y apoyar la toma de decisiones que mejoren la experiencia del cliente y el desempeño del negocio?
 
- **Objetivos específicos**
 
- Caracterizar los patrones de compra por categoría de producto, canal, dispositivo y ciudad de envío.
- Determinar qué variables se asocian con mayor probabilidad de devolución de un producto.
- Establecer la relación entre tiempo de entrega, devoluciones y calificación del cliente.

# Comprensión de los datos

- Llamada de liberías

In [1]:
import pandas as pd
import numpy as np

- Lectura Base De Datos

In [7]:
df=pd.read_csv("03_comercio_electronico_transacciones.csv")
df

,id_transaccion,fecha_compra,id_cliente,categoria_producto,precio_cop,metodo_pago,canal,dispositivo,ciudad_envio,tiempo_entrega_dias,calificacion_cliente_1a5,producto_devuelto
0,TXN-0005239,2022-12-28,14770.0,Hogar,159466.0,Tarjeta crédito,Sitio web,iOS,Pereira,5.0,5,No
1,TXN-0000262,2023-10-14,53229.0,Alimentos,72452.0,Tarjeta débito,Sitio web,Desktop,Cali,3.0,5,No
2,NaN,2022-05-04,19901.0,Hogar,162086.0,PSE,Sitio web,iOS,Cali,3.0,5,No
3,TXN-0005784,2024-11-05,49799.0,ALIMENTOS,41165.0,PSE,App móvil,Desktop,Bogotá,2.0,5,No
4,TXN-0001234,2022-05-16,31720.0,Tecnología,1138137.0,Tarjeta débito,Sitio web,Android,Bogotá,4.0,5,No
...,...,...,...,...,...,...,...,...,...,...,...,...
8075,TXN-0007097,2023-11-21,36456.0,Moda,124455.0,Contraentrega,App móvil,Android,Medellín,5.0,5,No
8076,TXN-0001668,2024-09-08,47592.0,Alimentos,41941.0,Contraentrega,App móvil,iOS,Bucaramanga,7.0,5,No
8077,TXN-0003322,2024-11-16,46273.0,Tecnología,1422437.0,Contraentrega,Sitio web,Android,Medellín,3.0,5,No
8078,TXN-0001689,2024-09-05,49760.0,Moda,104122.0,Tarjeta débito,App móvil,Desktop,Cartagena,6.0,5,No


- Información de la base de datos

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8080 entries, 0 to 8079
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id_transaccion            7835 non-null   object 
 1   fecha_compra              7825 non-null   object 
 2   id_cliente                7815 non-null   float64
 3   categoria_producto        8080 non-null   object 
 4   precio_cop                8080 non-null   float64
 5   metodo_pago               8080 non-null   object 
 6   canal                     8080 non-null   object 
 7   dispositivo               7826 non-null   object 
 8   ciudad_envio              8080 non-null   object 
 9   tiempo_entrega_dias       8080 non-null   float64
 10  calificacion_cliente_1a5  8080 non-null   int64  
 11  producto_devuelto         8080 non-null   object 
dtypes: float64(3), int64(1), object(8)
memory usage: 757.6+ KB


El conjunto de datos contiene 8.080 registros y 12 variables, de las cuales 8 son de tipo categórico es decir object, 3 numéricas continuas es decir float64 y 1 numérica discreta es decir int64. La mayoría de las variables no presentan valores faltantes; sin embargo, se identifican datos ausentes en las columnas id_transaccion, fecha_compra, id_cliente y dispositivo.

- Distribución y comportamiento de las variables numéricas

In [13]:
df.describe()

,id_cliente,precio_cop,tiempo_entrega_dias,calificacion_cliente_1a5
count,7815.000000,8.080000e+03,8080.000000,8080.000000
mean,34775.088420,2.485640e+05,4.257302,4.946287
std,14438.827505,3.071976e+05,1.925332,0.294499
min,10001.000000,5.000000e+03,1.000000,3.000000
25%,22033.500000,6.855125e+04,3.000000,5.000000
50%,34956.000000,1.201060e+05,4.000000,5.000000
75%,47089.000000,2.347180e+05,6.000000,5.000000
max,59991.000000,1.902012e+06,14.000000,5.000000


En este caso nos pudimos dar cuenta que las variables numéricas presentan diferentes comportamientos. Entre ellas pudimos ver que la variable id_cliente contiene 7.815 registros válidos, lo que confirma la presencia de 265 valores faltantes. Al tratarse de un identificador, sus medidas de tendencia central y dispersión no tienen interpretación estadística, ya que únicamente representan códigos de identificación.

La variable precio_cop presenta un precio promedio de $248.564 COP, con una mediana de $120.106 COP. La diferencia entre ambas medidas, junto con una desviación estándar de $307.198 COP, sugiere una distribución asimétrica hacia la derecha, influenciada por productos de alto valor. Los precios oscilan entre $5.000 COP y $1.902.012 COP, lo que evidencia una amplia variabilidad.

En cuanto al tiempo_entrega_dias, el promedio es de 4,26 días, con una mediana de 4 días y un rango entre 1 y 14 días. La desviación estándar de 1,93 días indica una variabilidad moderada en los tiempos de entrega.

Finalmente, la variable calificacion_cliente_1a5 muestra una media de 4,95 sobre 5 y una mediana de 5, con una desviación estándar de 0,29. Además, el 75 % de las observaciones corresponde a la calificación máxima (5), lo que indica que las valoraciones de los clientes están altamente concentradas en los valores más altos, sugiriendo una distribución sesgada hacia calificaciones positivas y una baja variabilidad.

- Columnas de la base de datos

In [11]:
df.columns

Index(['id_transaccion', 'fecha_compra', 'id_cliente', 'categoria_producto',
       'precio_cop', 'metodo_pago', 'canal', 'dispositivo', 'ciudad_envio',
       'tiempo_entrega_dias', 'calificacion_cliente_1a5', 'producto_devuelto'],
      dtype='object')

| Variable | Tipo de dato | Significado | Cómo se registra |
|---|---|---|---|
| `id_transaccion` | Texto (formato `TXN-#######`) | Identificador único de cada transacción o compra | Generado automáticamente por el sistema al momento de confirmarse la compra (llave primaria del registro) |
| `fecha_compra` | Fecha (AAAA-MM-DD) | Fecha en que se realizó la compra | Capturada por el sistema transaccional en el momento del checkout/pago |
| `id_cliente` | Numérico (almacenado como decimal, debería ser entero) | Identificador único del cliente que realiza la compra | Asignado al registrarse el cliente en la plataforma; se referencia en cada transacción que hace |
| `categoria_producto` | Categórica (texto) | Categoría a la que pertenece el producto comprado (Hogar, Alimentos, Tecnología, Moda, etc.) | Definida por el catálogo de productos de la plataforma; se hereda del producto al momento de la venta |
| `precio_cop` | Numérico (decimal) | Precio del producto/transacción en pesos colombianos | Tomado del catálogo/carrito de compras al momento de la transacción |
| `metodo_pago` | Categórica (texto) | Forma en que el cliente pagó (Tarjeta crédito, Tarjeta débito, PSE, Billetera digital, Contraentrega) | Capturado por la pasarela de pagos en el momento del checkout |
| `canal` | Categórica (texto) | Medio por el cual se hizo la compra (Sitio web, App móvil) | Registrado automáticamente por el sistema según el origen de la sesión |
| `dispositivo` | Categórica (texto) | Tipo de dispositivo usado (Android, iOS, Desktop) | Capturado por metadatos técnicos del navegador/app (user-agent) al momento de la sesión |
| `ciudad_envio` | Categórica (texto) | Ciudad destino del envío del pedido | Ingresada por el cliente (o tomada de su perfil) al definir la dirección de envío |
| `tiempo_entrega_dias` | Numérico (decimal) | Días transcurridos entre la compra y la entrega del producto | Calculado por el sistema logístico como diferencia entre fecha de compra y fecha real de entrega |
| `calificacion_cliente_1a5` | Numérico (entero, escala 1 a 5) | Calificación que el cliente da a la compra/experiencia | Ingresada manualmente por el cliente, típicamente en una encuesta post-compra |
| `producto_devuelto` | Categórica binaria (Sí/No) | Indica si el producto fue devuelto por el cliente | Registrado por el sistema de logística inversa cuando se procesa una devolución |

## Completitud de los datos

- Cantidad de nulos por variable

In [15]:
print(df.isnull().sum())

id_transaccion              245
fecha_compra                255
id_cliente                  265
categoria_producto            0
precio_cop                    0
metodo_pago                   0
canal                         0
dispositivo                 254
ciudad_envio                  0
tiempo_entrega_dias           0
calificacion_cliente_1a5      0
producto_devuelto             0
dtype: int64


In [16]:
print((df.isnull().sum()/len(df))*100)

id_transaccion              3.032178
fecha_compra                3.155941
id_cliente                  3.279703
categoria_producto          0.000000
precio_cop                  0.000000
metodo_pago                 0.000000
canal                       0.000000
dispositivo                 3.143564
ciudad_envio                0.000000
tiempo_entrega_dias         0.000000
calificacion_cliente_1a5    0.000000
producto_devuelto           0.000000
dtype: float64


El análisis de completitud de los 8.080 registros de la base de datos nos mostró un alto nivel de integridad, ya que la mayoría de las variables no presenta valores faltantes. Sin embargo, se identificaron cuatro variables con datos nulos, como por ejemplo lo podemos ver en  `id_transaccion` con 245 registros es decir en porcentaje 3,03%, `fecha_compra` con 255 registros es decir en porcentaje 3,16%, `id_cliente` con 265 registros es decir en porcentaje3,28% y `dispositivo` con 254 registros es decir en porcentaje 3,14%.

Aunque el porcentaje de valores faltantes es inferior al 5% en todos los casos, su impacto no es el mismo para todas las variables. En particular, la ausencia de información en `id_transaccion`, `id_cliente` y `fecha_compra` resulta relevante, ya que estas corresponden a identificadores y referencias temporales fundamentales para garantizar la trazabilidad de las transacciones, la correcta asociación con los clientes y el análisis cronológico de los eventos. En contraste, los valores faltantes en `dispositivo` tienen un impacto menor, dado que se trata de una variable descriptiva cuyo uso es principalmente analítico y cuya ausencia no impide identificar ni procesar las transacciones.


## Consistencia de los datos

In [17]:
# Revisar valores únicos de cada variable categórica
columnas_categoricas = [
    "categoria_producto", "metodo_pago", "canal",
    "dispositivo", "ciudad_envio", "producto_devuelto"
]
 
for col in columnas_categoricas:
    valores = sorted(df[col].dropna().unique())
    print(f"\n--- {col} ({len(valores)} valores únicos) ---")
    print(valores)
 


--- categoria_producto (20 valores únicos) ---
['ALIMENTOS', 'Alimentos', 'BELLEZA', 'Belleza', 'DEPORTES', 'Deportes', 'ELECTRÓNICA', 'Electrónica', 'HOGAR', 'Hogar', 'JUGUETES', 'Juguetes', 'LIBROS', 'Libros', 'MASCOTAS', 'MODA', 'Mascotas', 'Moda', 'TECNOLOGÍA', 'Tecnología']

--- metodo_pago (5 valores únicos) ---
['Billetera digital', 'Contraentrega', 'PSE', 'Tarjeta crédito', 'Tarjeta débito']

--- canal (2 valores únicos) ---
['App móvil', 'Sitio web']

--- dispositivo (3 valores únicos) ---
['Android', 'Desktop', 'iOS']

--- ciudad_envio (20 valores únicos) ---
['BARRANQUILLA', 'BOGOTÁ', 'BUCARAMANGA', 'Barranquilla', 'Bogotá', 'Bucaramanga', 'CALI', 'CARTAGENA', 'Cali', 'Cartagena', 'CÚCUTA', 'Cúcuta', 'IBAGUÉ', 'Ibagué', 'MANIZALES', 'MEDELLÍN', 'Manizales', 'Medellín', 'PEREIRA', 'Pereira']

--- producto_devuelto (2 valores únicos) ---
['No', 'Sí']


Se evaluó la consistencia de las seis variables categóricas de la base de datos mediante el análisis de sus valores únicos. En la variable `metodo_pago` pudimos encontrar Billetera digital, Contraentrega, PSE, Tarjeta crédito y Tarjeta débito, `canal` encontramos App móvil y Sitio web, `dispositivo` encontramos Android, Desktop e iOS y `producto_devuelto` encontramos Sí y No presentan valores homogéneos y sin inconsistencias de formato, lo que garantiza su correcta utilización en análisis descriptivos y modelos posteriores.

Por otro lado, las variables `categoria_producto` y `ciudad_envio` registran 20 valores únicos cada una, aunque conceptualmente corresponden a únicamente 10 categorías de producto y 10 ciudades. Esta diferencia se debe a inconsistencias en el formato del texto, ya que un mismo valor fue almacenado tanto en mayúscula sostenida como en formato título. Como por ejemplo, `ALIMENTOS` y `Alimentos`, o `CÚCUTA` y `Cúcuta`. Si bien esta situación no altera el significado de los datos, sí afecta su consistencia, provocando que una misma categoría sea interpretada como dos valores distintos durante los procesos de agrupación, conteo o visualización. En consecuencia, es necesario normalizar previamente el formato de estas variables.

In [18]:
# Detectar específicamente duplicados por formato de texto
def revisar_duplicados_formato(columna):
    reales = df[columna].dropna().str.upper().nunique()
    encontrados = df[columna].dropna().nunique()
    print(f"{columna}: {encontrados} valores únicos encontrados vs "
          f"{reales} categorías reales (ignorando mayúsculas)")
 
print("\nComparación de valores únicos vs categorías reales:")
revisar_duplicados_formato("categoria_producto")
revisar_duplicados_formato("ciudad_envio")


Comparación de valores únicos vs categorías reales:
categoria_producto: 20 valores únicos encontrados vs 10 categorías reales (ignorando mayúsculas)
ciudad_envio: 20 valores únicos encontrados vs 10 categorías reales (ignorando mayúsculas)


Al comparar la cantidad de valores únicos con y sin distinción de mayúsculas/minúsculas, se confirmó que las inconsistencias detectadas en categoria_producto y ciudad_envio corresponden en su totalidad a duplicados por formato de texto y no a categorías o ciudades realmente distintas. En ambos casos, los 20 valores únicos encontrados se reducen exactamente a 10 al ignorar la diferencia entre mayúsculas y minúsculas, lo que indica que cada una de las 10 categorías de producto y cada una de las 10 ciudades de envío está registrada bajo dos formatos distintos dentro del mismo conjunto de datos. Este hallazgo confirma que no se trata de errores de captura de nuevos valores, sino de una falta de estandarización en el formato de texto al momento del registro.

In [19]:
# Revisar tipo de dato de id_cliente (debería ser entero, no decimal)
no_enteros = df["id_cliente"].dropna().apply(lambda x: x != int(x)).sum()
print(f"\nid_cliente: tipo actual = {df['id_cliente'].dtype}, "
      f"valores no enteros = {no_enteros}")


id_cliente: tipo actual = float64, valores no enteros = 0


Se verificó el tipo de dato de la variable id_cliente, encontrándose que está almacenada como número decimal (float64) en lugar de un identificador entero. Al revisar si esta representación afecta la integridad de los valores, se confirmó que los 7.815 registros no nulos corresponden efectivamente a números enteros (0 valores con decimales reales), por lo que el problema es exclusivamente de formato de almacenamiento y no de calidad del dato en sí. Esta inconsistencia de tipo probablemente se origina en la presencia de valores nulos en la columna, ya que pandas convierte automáticamente a decimal cualquier columna numérica que contenga valores faltantes.

In [20]:
# Revisar rango de calificacion_cliente_1a5
print("\nValores presentes en calificacion_cliente_1a5:")
print(sorted(df["calificacion_cliente_1a5"].unique()))
print(df["calificacion_cliente_1a5"].value_counts().sort_index())


Valores presentes en calificacion_cliente_1a5:
[np.int64(3), np.int64(4), np.int64(5)]
calificacion_cliente_1a5
3     145
4     144
5    7791
Name: count, dtype: int64


Se examinó el rango de valores de calificacion_cliente_1a5, encontrándose que, si bien la escala nominal del campo es de 1 a 5, en la práctica los datos solo contienen los valores 3, 4 y 5, sin registro alguno de calificaciones 1 o 2. Adicionalmente, se identificó una fuerte concentración en el valor máximo con un valor de 7.791 de los 8.080 registros presentan calificación 5, mientras que únicamente 289 registros tienen calificación 3 o 4. Si bien esto no constituye un error de calidad de datos en sí mismo, evidencia una baja variabilidad en la variable objetivo del análisis de satisfacción, lo cual debe considerarse al momento de interpretar o modelar los factores asociados a una menor calificación.

## Trazabilidad de los datos